ui/src/app/core/capabilities-activator.service.ts

```diff
@@ -0,0 +1,76 @@
+import {ActivatedRouteSnapshot, RouterStateSnapshot, Router} from '@angular/router';
+import {CanActivate} from '@angular/router';
+import {Injectable} from '@angular/core';
+
+import {AuthService} from './auth.service';
+import {CapabilitiesResponse} from '../shared/model/CapabilitiesResponse';
+import {CapabilitiesService} from './capabilities.service';
+import {environment} from '../../environments/environment';
+import {URLSearchParamsUtils} from '../shared/utils/url-search-params.utils'
+
+/** Service wrapper for routing based on current authentication state. */
+@Injectable()
+export class CapabilitiesActivator implements CanActivate {
+
+  constructor(
+    private readonly authService: AuthService,
+    private readonly capabilitiesService: CapabilitiesService,
+    private readonly router: Router) {}
+
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+    return this.capabilitiesService.getCapabilities()
+      .then(cap => this.handleAuthCapabilities(cap, route.routeConfig.path, state.url))
+      .then(cap => this.handleProjectCapabilities(cap, route))
+      .catch(reason => {
+          // Handle all not-activated errors by just returning false for
+          // this promise. All others, re-throw them.
+          if (reason == 'not-activated') {
+            return false;
+          }
+          throw reason;
+        });
+  }
+
+  private handleAuthCapabilities(capabilities: CapabilitiesResponse, path: String, url: String): Promise<CapabilitiesResponse> {
```

Comment:
A `then` accepts a callback which can return a value, or a `Promise`; in this case since nothing async is happening, it would make more sense to just return the value directly, and throw to trigger the error path.

Reply:
This function actually does make an async call to `this.authService.isAuthenticated()`. But I changed the function to return `Promise<CapabilitiesResponse>|CapabilitiesResponse` so that we can return directly in cases where the async call is not made, and throw errors instead of rejecting the promise.

I changed it to return the value directly for `handleProjectCapabilities` because that does nothing async.